pkg/agent/openflow/pipeline.go

```diff
@@ -454,388 +502,525 @@ func (c *client) DeleteOFEntries(ofEntries []binding.OFEntry) error {
+			// PipelineClassifierTable.
+			flows = append(flows, pipelineClassifyFlow(cookieID, binding.ProtocolARP, pipeline))
+		case pipelineMulticast:
+			// This generates the flow to match multicast packets and forward them to the first table of pipeline for
+			// multicast in IPClassifierTable. Note that, IPClassifierTable is in stageValidation of pipeline for IP.
+			// In another word, pipeline for multicast is forked from IPClassifierTable in pipeline for IP.
+			flows = append(flows, multicastPipelineClassifyFlow(cookieID, pipeline))
 		}
-		flows = append(flows, flowBuilder.Cookie(c.cookieAllocator.Request(cookie.Default).Raw()).Done())
 	}
+	// This generates default the flow for PipelineClassifierTable.
+	flows = append(flows, PipelineClassifierTable.ofTable.BuildFlow(priorityMiss).
+		Cookie(cookieID).
+		Action().Drop().
+		Done())
+
 	return flows
 }
 
-// tunnelClassifierFlow generates the flow to mark traffic comes from the tunnelOFPort.
-func (c *client) tunnelClassifierFlow(tunnelOFPort uint32, category cookie.Category) binding.Flow {
-	nextTable := ConntrackTable
-	if c.proxyAll {
-		nextTable = ServiceConntrackTable
-	}
-	return ClassifierTable.BuildFlow(priorityNormal).
+// Feature: featurePodConnectivity
+// Stage: stageClassifier
+// Tables: ClassifierTable
+// Refactored from:
+//   - func (c *client) tunnelClassifierFlow(tunnelOFPort uint32, category cookie.Category) binding.Flow
+// tunnelClassifierFlow generates the flow to mark the packets from tunnel port.
+func (f *featurePodConnectivity) tunnelClassifierFlow(tunnelOFPort uint32) binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityNormal).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
 		MatchInPort(tunnelOFPort).
 		Action().LoadRegMark(FromTunnelRegMark).
 		Action().LoadRegMark(RewriteMACRegMark).
-		Action().GotoTable(nextTable.GetID()).
-		Cookie(c.cookieAllocator.Request(category).Raw()).
+		Action().GotoStage(stageConntrackState).
 		Done()
 }
 
-// gatewayClassifierFlow generates the flow to mark traffic comes from the gatewayOFPort.
-func (c *client) gatewayClassifierFlow(category cookie.Category) binding.Flow {
-	return ClassifierTable.BuildFlow(priorityNormal).
+// Feature: featurePodConnectivity
+// Stage: stageClassifier
+// Tables: ClassifierTable
+// Refactored from:
+//   - func (c *client) gatewayClassifierFlow(category cookie.Category) binding.Flow
+// gatewayClassifierFlow generates the flow to mark the packets from Antrea gateway port.
+func (f *featurePodConnectivity) gatewayClassifierFlow() binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityNormal).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
 		MatchInPort(config.HostGatewayOFPort).
 		Action().LoadRegMark(FromGatewayRegMark).
-		Action().GotoTable(ClassifierTable.GetNext()).
-		Cookie(c.cookieAllocator.Request(category).Raw()).
+		Action().GotoStage(stageValidation).
```

Comment:
In the framework, is it preferred to use `NextTable` or specify stage?

Reply:
Yes. `NextTable`, `GotoStage` and `ResubmitToTables` is recommended.